github/codeowners.py

```diff
@@ -81,73 +111,173 @@ def filter_codeowners_entries(
+) -> typing.Generator[github3.users.ShortUser, None, None]:
+    '''
+    Returns generator yielding members from team.
+    Absent teams passed silently with `absent_ok`.
+    Subteams are resolved as well.
+    '''
     try:
-        return next(iterable)
-    except StopIteration:
-        return None
+        team = org.team_by_name(team_name)
+    except github3.exceptions.NotFoundError:
+        if absent_ok:
+            return
+        raise
+    # resolves subteams as well
+    yield from team.members()
 
 
-def determine_email_address(
-    github_user_name: str,
-    github_api: GitHub,
-) -> typing.Optional[str]:
-    not_none(github_user_name)
+def mail_address_for_user(
+    user_name: str,
+    gh_api: GitHub | GitHubEnterprise,
+    absent_ok: bool=True,
+) -> str | None:
+    '''
+    Returns email associated with github user for username.
+    `None` returned if user not found and `absent_ok`.
+    Users without associated email return `None`.
+    '''
     try:
-        user = github_api.user(github_user_name)
+        user = gh_api.user(user_name)
     except NotFoundError:
-        logger.warning(f'failed to lookup {github_user_name=} {github_api._github_url=}')
-        return None
+        if absent_ok:
+            logger.warning(f'failed to lookup {user_name=} {gh_api._github_url=}')
+            return None
+        raise
 
     return user.email
 
 
-def resolve_team_members(
-    github_team_name: str,
-    github_api: GitHub,
-) -> typing.Union[typing.Generator[str, None, None], list]:
-    not_none(github_team_name)
-    org_name, team_name = github_team_name.split('/') # always of form 'org/name'
-    organisation = github_api.organization(org_name)
-    # unfortunately, we have to look-up the team (no api to retrieve it by name)
-    team_or_none = _first(filter(lambda team: team.slug == team_name, organisation.teams()))
-    if not team_or_none:
-        logger.warning('failed to lookup team {t}'.format(t=team_name))
-        return []
-    for member in map(github_api.user, team_or_none.members()):
-        if member.email:
-            yield member.email
-        else:
-            logger.warning(f'no email found for GitHub user {member}')
+def user_for_mail_address(
+    mail_address: str,
+    gh_api: GitHub | GitHubEnterprise,
+) -> github3.users.ShortUser | None:
+    '''
+    Find user for email address.
+    Exactly one user for `None` will be returned.
+    '''
+    results = list(gh_api.search_users(query=f'{mail_address} in:email'))
+    results: list[github3.search.user.UserSearchResult]
 
+    if not results:
+        return None
 
-def resolve_email_addresses(
-    codeowners_entries,
-    github_api: GitHub,
-) -> typing.Generator[str, None, None]:
+    if len(results) > 1:
+        logger.warning(f'more than one user found for {mail_address=}')
+        return None
+
+    return results[0].user
+
+
+def iter_codeowners(
+    codeowners_entries: collections.abc.Collection[str],
```

Comment:
nit: if we do not need to iterate more than once, it is more adequate to expect an typing.Iterable
also, why not pass-in `Codeowner`-objects?

the function's name is misleading. while it does return a generator, what it actually does is to convert `str` into typed objects (Username, Teamname, Emailaddress).

This should reasonably already be done at the function(s) we have that currently emit entries from CODEOWNERS files